### PR TITLE
[bls-signatures] Initial integration

### PR DESCRIPTION
**Pull request by guidovranken**
Differential fuzzing of 3 BLS signature libraries

- blst https://github.com/supranational/blst
- Chia bls-signatures https://github.com/Chia-Network/bls-signatures/
- Herumi mcl https://github.com/herumi/mcl

BLS signatures are a critical security feature in cryptocurrency clients, such as:

- Ethereum 2.0
- Chia
- DFinity
- Filecoin

If this project is accepted, I will add more auto_cc's later.

### CODE REVIEW

**Comment by AdamKorcz** [comment]
Could you adjust the title of the PR to suggest this is an initial integration?

File: projects/bls-signatures/build.sh [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+# Build mcl
+cd $SRC/mcl/
+mkdir build/
+cd build/
+if [[ $CFLAGS != *-m32* ]]
+then
+    cmake .. -DMCL_STATIC_LIB=on 
+    export LINK_FLAGS="$LINK_FLAGS -lgmp"
+else
+    cmake .. -DMCL_STATIC_LIB=on -DGMP_INCLUDE_DIR="$SRC/libgmp-install/include/" -DGMP_LIBRARY="$SRC/libgmp-install/lib/libgmp.a" -DGMP_GMPXX_INCLUDE_DIR="$SRC/libgmp-install/include/" -DGMP_GMPXX_LIBRARY="$SRC/libgmp-install/lib/libgmpxx.a" -DMCL_USE_ASM=off
```
Comment (AdamKorcz) [inline projects/bls-signatures/build.sh:94]: This looks fine in the terminal, but for Github, could we shorten the line?

```bash
cmake .. -DMCL_STATIC_LIB=on \
    -DGMP_INCLUDE_DIR="$SRC/libgmp-install/include/" 
    -DGMP_LIBRARY="$SRC/libgmp-install/lib/libgmp.a" 
    -DGMP_GMPXX_INCLUDE_DIR="$SRC/libgmp-install/include/" 
    -DGMP_GMPXX_LIBRARY="$SRC/libgmp-install/lib/libgmpxx.a" 
    -DMCL_USE_ASM=off
```

File: projects/bls-signatures/build.sh [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+export MCL_INCLUDE_PATH=$(realpath ../include/)
+export MCL_LIBMCL_A_PATH=$(realpath lib/libmcl.a)
+export MCL_LIBMCLBN384_A_PATH=$(realpath lib/libmclbn384.a)
+export CXXFLAGS="$CXXFLAGS -DCRYPTOFUZZ_MCL"
+
+# Build Botan
+cd $SRC/botan/
+if [[ $CFLAGS != *-m32* ]]
+then
+    ./configure.py --cc-bin=$CXX --cc-abi-flags="$CXXFLAGS" --disable-shared --disable-modules=locking_allocator,x509,tls --build-targets=static --without-documentation
```
Comment (AdamKorcz) [inline projects/bls-signatures/build.sh:107]: same here:

```bash
./configure.py --cc-bin=$CXX \
    --cc-abi-flags="$CXXFLAGS" \
    --disable-shared \
    --disable-modules=locking_allocator,x509,tls \
    --build-targets=static \
    --without-documentation
```

File: projects/bls-signatures/build.sh [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+else
+    ./configure.py --cpu=x86_32 --cc-bin=$CXX --cc-abi-flags="$CXXFLAGS" --disable-shared --disable-modules=locking_allocator,x509,tls --build-targets=static --without-documentation
```
Comment (AdamKorcz) [inline projects/bls-signatures/build.sh:109]: Same here:

```bash
./configure.py --cpu=x86_32 \
    --cc-bin=$CXX \
    --cc-abi-flags="$CXXFLAGS" \
    --disable-shared \
    --disable-modules=locking_allocator,x509,tls \
    --build-targets=static \
    --without-documentation
```

File: projects/bls-signatures/project.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+primary_contact: "guidovranken@gmail.com"
+main_repo: "https://github.com/supranational/blst.git"
+sanitizers:
+ - address
+ - undefined
+architectures:
+ - x86_64
+ - i386
+auto_ccs:
+    - "justin@ethereum.org"
```
Comment (AdamKorcz) [inline projects/bls-signatures/project.yaml:12]: A small thing: Same indentation throughout the entire `project.yaml`

File: projects/bls-signatures/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM gcr.io/oss-fuzz-base/base-builder
+RUN apt-get update && apt-get install -y make autoconf automake libtool cmake wget python lzip libgmp-dev
```
Comment (AdamKorcz) [inline projects/bls-signatures/Dockerfile:18]: It should not be necessary to install `cmake`.

**Comment by guidovranken** [comment]
@AdamKorcz I've addressed the issues you raised.

The CI says the UBSAN build fails the build check, but I cannot reproduce this locally.

**Comment by AdamKorcz** [comment]
I tried this locally. The broken build reproduces with the following steps:

1. `sudo python3 infra/helper.py pull_images`
2. `sudo python3 infra/helper.py build_image bls-signatures`
3. `sudo python3 infra/helper.py build_fuzzers bls-signatures --engine libfuzzer --sanitizer undefined --architecture x86_64`
4. `sudo python3 infra/helper.py check_build bls-signatures --engine libfuzzer --sanitizer undefined --architecture x86_64` <- Build error happens here.

**Comment by AdamKorcz** [comment]
I followed the steps taken in my last comment but with the build file from the first commit in this PR (which passed CI), and the same errors were encountered during the `check_build` step.

**Comment by guidovranken** [comment]
> I tried this locally. The broken build reproduces with the following steps:
> 
>     1. `sudo python3 infra/helper.py pull_images`
> 
>     2. `sudo python3 infra/helper.py build_image bls-signatures`
> 
>     3. `sudo python3 infra/helper.py build_fuzzers bls-signatures --engine libfuzzer --sanitizer undefined --architecture x86_64`
> 
>     4. `sudo python3 infra/helper.py check_build bls-signatures --engine libfuzzer --sanitizer undefined --architecture x86_64` <- Build error happens here.

I repeated your steps verbatim (after doing a `docker system prune -a`) and cannot reproduce. I've disabled the UBSAN build for now.

**Comment by AdamKorcz** [comment]
Great. This one is ready for review @oliverchang @jonathanmetzman @inferno-chromium @asraa

**Review by jonathanmetzman — APPROVED** [review]
LGTM

**Pull request merged**